### PR TITLE
DO NOT MERGE: regex tests

### LINKED ISSUE #7: Experiment on data processing (regex)

### PR DESCRIPTION
**Pull request by Claiyc**
# PR description
some regex testing with 100% matches.

This may be a valid general approach for matching the value regex considering constraints.

Next up: accept values (and constraints) that only match to a certain percentage. See `Levenshtein Algorithm`.
References:
- https://stackoverflow.com/questions/40010113/how-can-i-find-the-percentage-of-a-regex-match-on-a-string
- https://stackoverflow.com/questions/10473745/compare-strings-javascript-return-of-likely
- https://www.npmjs.com/package/string-similarity

# Definition Of Done (DoD)
*This PR can be squashed / merged if*
- [ ] a developer is assigned
- [ ] the PR is NOT estimated
- [ ] the PR is labeled
- [ ] the PR is assigned to the current sprint
- [ ] a meaningful title has been set according to https://www.conventionalcommits.org/
- [ ] the PR is described in detail
- [ ] the PR links to an issue
- [ ] the PR has been reviewed

*Add additional conditions here if necessary for this PR*
fixes #7

### CODE REVIEW

**Comment by Claiyc** [comment]
Research, DO NOT MERGE! Feel free to leave your reviews anyway.

File: src/proc/Vigad.ts [hunk 1]
```diff
@@ -1,5 +1,28 @@
-class Vigad {
+export class Vigad {
```
Comment (BeierKevin) [inline src/proc/Vigad.ts:1]: Maybe a good documentation with JSDoc style would be great for the future.

**Comment by Claiyc** [comment]
getting the percentage of how good a string matches a regex seems to be quite the complex problem: https://stackoverflow.com/questions/23634829/similarity-between-pattern-regex-and-the-value-found

An approach may also be to generate a (large) set of strings matching the regex
(https://stackoverflow.com/questions/21555400/generating-strings-from-regular-expression-in-javascript ) and test those against the value found (using levenshteins algorithm or something implemented like https://www.npmjs.com/package/string-similarity )

**Comment by Claiyc** [comment]
https://gist.github.com/lopespm/53a215d0b2b0518b52b6bb6687bdaff6
this looks promising as well (python)

**Comment by Claiyc** [comment]
additional ideas / references:
- https://laurikari.net/tre/
- http://www.levenshtein.net/levenshtein_search.htm
- https://www.quora.com/How-do-you-find-out-the-edit-distance-between-a-given-string-and-a-RegExp

**Comment by Claiyc** [comment]
> https://gist.github.com/lopespm/53a215d0b2b0518b52b6bb6687bdaff6 this looks promising as well (python)

this only works for very specific cases

**Comment by Claiyc** [comment]
Alternative to https://github.com/VisualGameData/VIGAD/pull/21#issuecomment-1324387842

(works with relatively specific regex) Generate all valid words with edit distance 0 -> array.is_inside(word)
=> Generate all valid words with e-d 1 -> array.is_inside(word)
...

**Comment by Claiyc** [comment]
grafik
This may be useful later on

**Comment by Claiyc** [comment]
https://github.com/codebox/homoglyph may be useful

**Comment by Claiyc** [comment]
https://github.com/nodeca/unhomoglyph even better

**Comment by Claiyc** [comment]
Worth mentioning: tesseract can be configured to only search for specific characters: https://stackoverflow.com/questions/2363490/limit-characters-tesseract-is-looking-for

**Review by roman533 — DISMISSED** [review]

**Comment by Claiyc** [comment]
I'm relatively happy with this now. Feel free to take a look, it's quite complicated though.
The current approach is now mainly based on the generation of (many) matches with levenshtein distance 0 (for every regex) and comparison of these matches with the scanned text using https://www.npmjs.com/package/string-similarity#findbestmatchmainstring-targetstrings (based on [Dice's Coefficient](http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient -> alternative to levenshtein algorithm)

**Comment by Claiyc** [comment]
Currently I plan on having multiple options for the user to choose from for each regex (documented in file).
These are
- testing generated matches against ALL substrings of scanned text (or against substrings split at spaces)
- Approximate value matching or exact value matching
- number of matches to be generated (current tests show that a good default-amount is ~50000)

**Comment by Claiyc** [comment]
Still to test:
- [ ] similar letter-number conversion (and vise versa). See https://github.com/VisualGameData/VIGAD/pull/21#issuecomment-1324808310

**Comment by Claiyc** [comment]
So... I just talked to an expert on this matter and he basically said that what we have right now is bullshit 😂 I mean yes it works and yes it may actually be used in production BUT it still is an incredible overhead... "approximate" regex matching can be done MUCH MUCH more efficient with a bit more complex regex.

Will test around a bit...

**Comment by Claiyc** [comment]
closing as completed (see #7 )

**Comment by Claiyc** [comment]
(this branch should not be deleted)

**Pull request closed without merging**